src/Storages/StorageReplicatedMergeTree.h

```diff
@@ -447,8 +450,21 @@ class StorageReplicatedMergeTree final : public MergeTreeData
     /// A thread that processes reconnection to ZooKeeper when the session expires.
     ReplicatedMergeTreeRestartingThread restarting_thread;
 
+    ReplicatedMergeTreeAttachThread attach_thread;
+
     PartMovesBetweenShardsOrchestrator part_moves_between_shards_orchestrator;
 
+    enum class InitializationPhase : uint8_t
+    {
+        INITIALIZING,
+        INITIALIZATION_DONE,
+        STARTUP_IN_PROGRESS,
+        STARTUP_DONE
+    };
+    mutable std::mutex initialization_mutex;
+    TSA_GUARDED_BY(initialization_mutex) InitializationPhase init_phase{InitializationPhase::INITIALIZING};
+    TSA_GUARDED_BY(initialization_mutex) bool startup_called{false};
```

Comment:
Logic with 4 states and additional flag looks quite complex to understand, please add a comment describing these state (in combination with flag values) and possible transitions between them